### DISCUSSION TITLE
Valhalla 3.1.0 and later is incompatible with Lua 5.2 or later

### DISCUSSION THREAD

**Discussion by mloskot** [valhalla/valhalla#3207] [Show and tell]
This is not a bug report but sharing an experience with building Valhalla 3.1.2 using plain Lua 5.2 instead of LuaJIT

## Lua to LuaJIT

Valhalla switched to LuaJIT in  https://github.com/valhalla/valhalla/pull/2352 and the change was released with Valhalla 3.1.0

Lua vs LuaJIT apparently have syntax differences https://github.com/valhalla/valhalla/pull/2521#issuecomment-671429428

## Trying Lua 5.2

I replaced LuaJIT with Lua 5.2.4 in my build of Valhalla and looks like I hit the Lua syntax incompatibilities while running `valhalla_build_tiles` as the log is full of errors (below) and the generated tiles are incomplete as `valhalla_run_route` finds nothing really.

```
2021/07/08 17:19:58.594394 [ERROR] Exception in Lua function: ways_proc: Failed to execute lua function for basic tag processing in way 3161851: [string "--TODO: check if you can use lua type boolean..."]:1882: attempt to index global 'bit' (a nil value)
...
2021/07/08 17:20:17.167196 [ERROR] Exception in Lua function: nodes_proc: Failed to execute lua function for basic tag processing in node 0: [string "--TODO: check if you can use lua type boolean..."]:1882: attempt to index global 'bit' (a nil value)
...
2021/07/08 17:20:18.545371 [ERROR] Exception in Lua function: rels_proc: Failed to execute lua function for basic tag processing in relation 12461: [string "--TODO: check if you can use lua type boolean..."]:1882: attempt to index global 'bit' (a nil value)
```

## Back to LuaJIT 2.0.5

As soon as I build Valhalla with LuaJIT 2.0.5, the errors are gone, tiles are generated and routing works fine.

## Conclusion

Valhalla 3.1.0 or later is not compatible with syntax of Lua 5.2 or later.

**Comment by purew** [comment]
This is intentional. LuaJit implements Lua 5.1 syntax which is incompatible with 5.2. We switched to LuaJit because of the performance benefits.

**Reply by mloskot** [reply]
Yes, I've learned about the reasons from the PR/comments linked above.

I just was trying my luck and I thought I'll share the results in case anyone else is going to try similar.
Besides, exact Lua version requirements are not displayed very vividly.
